gio/src/list_model.rs

```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if self.item_type() != LT::static_type() {
+            return Err(self);
+        }
+        let rc = Rc::new(Cell::new(false));
+
+        let rcc = rc.clone();
+        let signal_id = Cell::new(Some(self.connect_items_changed(move |_, _, _, _| {
+            rcc.replace(true);
+        })));
+
+        Ok(ListModelIter {
+            ty: Default::default(),
+            i: 0,
+            model: self.clone().upcast(),
+            changed: rc,
+            signal_id,
+        })
+    }
+}
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+#[error("the list model was mutated during iteration")]
+pub struct ListModelMutatedDuringIter;
+
+pub struct ListModelIter<T: IsA<glib::Object>> {
+    ty: PhantomData<T>,
+    i: u32,
+    model: ListModel,
+    changed: Rc<Cell<bool>>,
+    signal_id: Cell<Option<SignalHandlerId>>,
+}
+impl<T: IsA<glib::Object>> Iterator for ListModelIter<T> {
+    type Item = Result<T, ListModelMutatedDuringIter>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.changed.get() {
```

Comment:
You should get the number of items in the model when creating the iterator and then work based on that. The number of items can't change duration iteration, and by doing that you can implement
  - `Iterator::size_hint()` and `ExactSizeIterator`
  - `FusedIterator`
  - `DoubleEndedIterator`